gnovm/pkg/gnolang/coverage.go

```diff
@@ -0,0 +1,642 @@
+	Enabled        bool // -cover flag activated
+	PkgPath        string
+	rootDir        string
+	CurrentPackage string
+	currentFile    string
+	pathCache      pathCache
+	Files          fileCoverageMap
+	mu             sync.RWMutex
+}
+
+// FileCoverage stores coverage information for a single file
+type FileCoverage struct {
+	TotalLines      int
+	HitLines        map[int]int
+	ExecutableLines map[int]bool
+}
+
+func NewCoverageData(rootDir string) *CoverageData {
+	return &CoverageData{
+		rootDir:        rootDir,
+		PkgPath:        "",
+		CurrentPackage: "",
+		currentFile:    "",
+		Files:          make(fileCoverageMap),
+		pathCache:      make(pathCache),
+	}
+}
+
+func (c *CoverageData) Enable()         { c.Enabled = true }
+func (c *CoverageData) Disable()        { c.Enabled = false }
+func (c *CoverageData) IsEnabled() bool { return c.Enabled }
+
+// setExecutableLines sets the executable lines for a given file path in the coverage data.
+// It updates the ExecutableLines map for the given file path with the provided executable lines.
+func (c *CoverageData) setExecutableLines(filePath string, executableLines map[int]bool) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	cov, exists := c.Files[filePath]
+	if !exists {
+		cov = FileCoverage{
+			TotalLines:      0,
+			HitLines:        make(map[int]int),
+			ExecutableLines: make(map[int]bool),
+		}
+	}
+
+	cov.ExecutableLines = executableLines
+	c.Files[filePath] = cov
+}
+
+// updateHit updates the hit count for a given line in the coverage data.
+// This function is used to update the hit count for a specific line in the coverage data.
+// It increments the hit count for the given line in the HitLines map for the specified file path.
+func (c *CoverageData) updateHit(pkgPath string, line int) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	if !c.isValidFile(pkgPath) {
+		return
+	}
+
+	fileCoverage := c.getOrCreateFileCoverage(pkgPath)
+
+	if fileCoverage.ExecutableLines[line] {
+		fileCoverage.HitLines[line]++
+		c.Files[pkgPath] = fileCoverage
+	}
+}
+
+func (c *CoverageData) isValidFile(pkgPath string) bool {
+	return strings.HasPrefix(pkgPath, c.PkgPath) &&
+		strings.HasSuffix(pkgPath, ".gno") &&
+		!isTestFile(pkgPath)
+}
+
+func (c *CoverageData) getOrCreateFileCoverage(pkgPath string) FileCoverage {
+	fileCoverage, exists := c.Files[pkgPath]
+	if !exists {
+		fileCoverage = FileCoverage{
+			TotalLines: 0,
+			HitLines:   make(map[int]int),
+		}
+		c.Files[pkgPath] = fileCoverage
+	}
+	return fileCoverage
+}
+
+func (c *CoverageData) addFile(filePath string, totalLines int) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	if isTestFile(filePath) {
+		return
+	}
+
+	fileCoverage, exists := c.Files[filePath]
+	if !exists {
+		fileCoverage = FileCoverage{
+			HitLines: make(map[int]int),
+		}
+	}
+
+	fileCoverage.TotalLines = totalLines
+	c.Files[filePath] = fileCoverage
+}
+
+// Report prints the coverage report to the console
+func (c *CoverageData) Report(io commands.IO) {
```

Comment:
Having `commands.IO` here, instead of a regular `io.WriteCloser` is heavy implementation leaking, especially in the VM

Reply:
Same goes for `ViewFiles`. The VM package shouldn't import `commands.IO`